src/internal/hooks/use-latency-metrics/index.ts

```diff
@@ -0,0 +1,125 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+import { useEffect, useRef } from 'react';
+import { useIdFallback } from '../use-unique-id';
+import { useEffectOnUpdate } from '../use-effect-on-update';
+import { isInViewport } from './is-in-viewport';
+import { metrics } from '../../analytics/metrics';
+
+export function useLatencyMetrics(
+  componentName: string,
+  elementRef: React.RefObject<HTMLElement>,
+  instanceId: string | undefined,
+  loading: boolean | undefined = false,
+  componentType: 'spinner' | undefined = undefined
+) {
+  const lifecycleId = useIdFallback();
+
+  const isLoadingOrSpinner = loading || componentType === 'spinner';
+
+  const loadingStartTime = useRef<undefined | number>(undefined);
+  if (isLoadingOrSpinner && loadingStartTime.current === undefined) {
+    loadingStartTime.current = performance.now();
+  }
+
+  useEffect(() => {
+    if (!elementRef.current) {
+      return;
+    }
+
+    const timestamp = Date.now();
+
+    const cleanup = isInViewport(elementRef.current!, inViewport => {
+      emitMetric(
+        {
+          type: 'mounted',
+          lifecycleId,
+          componentName,
+          inViewport,
+          metadata: { instanceId },
+          loading: isLoadingOrSpinner,
+          loadingDuration: undefined,
+        },
+        timestamp
+      );
+    });
+
+    return () => {
+      cleanup();
+
+      if (loadingStartTime.current !== undefined) {
+        const loadingDuration = performance.now() - loadingStartTime.current;
+
+        emitMetric(
+          {
+            type: componentType === 'spinner' ? 'loading-finished' : 'loading-cancelled',
+            lifecycleId,
+            componentName,
+            inViewport: undefined,
+            metadata: { instanceId },
+            loading: isLoadingOrSpinner,
+            loadingDuration,
+          },
+          Date.now()
+        );
+      }
+    };
+    // eslint-disable-next-line react-hooks/exhaustive-deps
```

Comment:
I agree, I'll try to improve it with ref objects.

I'm hesitant to use a _callback_ ref, since it seems semantically different from when an Effect runs. For example, if/when React uses [its proposed Offscreen rendering](https://react.dev/blog/2023/03/22/react-labs-what-we-have-been-working-on-march-2023#offscreen-rendering), the lifecycle of Effects could be separate from the one of callback refs.